### ISSUE TITLE
Calculate average for overlapping parts when merging rasters

### ISSUE THREAD

**Issue by magnusnil** [OSGeo/gdal#5176]
I tried using the Merge tool in QGIS 3.16.10 (based on GDAL 3.1.4) to merge a few raster files. The files overlap with a few hundred meters and it would be very useful if GDAL could calculate the average value for the raster files (2 or more files) where they overlap. See for example how FME allows this kind of manipulation, http://docs.safe.com/fme/html/FME_Desktop_Documentation/FME_Transformers/Transformers/rastermosaicker.htm.

QGIS issue: https://github.com/qgis/QGIS/issues/46990

**Comment by rouault** [comment]
You can probably achieve that with gdalwarp -r average.  gdal_merge.py is a simple/simplistic tool

**Comment by jratike80** [comment]
I think that gdalwarp does not do it or I just do not know how. Simple test with overlapping red (1) and green (2) images

```
gdal_create -outsize 100 100 -a_ullr 0 100 100 0 -bands 3 -burn 255 -burn 0 -burn 0 img1.tif
gdal_create -outsize 100 100 -a_ullr 50 100 150 0 -bands 3 -burn 0 -burn 255 -burn 0 img2.tif

gdalwarp -r average img2.tif img1.tif warped.tif
```

<img width="120" alt="image" src="https://user-images.githubusercontent.com/1751612/151156421-754bbc05-faaa-4c54-b8a4-81e026957e44.png">

The latter image, red, covers the first one. There was an initiative that should be perfect for this kind of tasks https://gdal.org/development/rfc/rfc62_raster_algebra.html, if implemented.

**Comment by rouault** [comment]
you're right. The resampling method in gdalwarp is about how to resample source pixels from a single source. For the purpose you want to accmoplish you'll need to use several steps:
- compute the overlapping area
- use gdal_translate/gdalwarp to extract from each source the overlapping area
- use gdal_calc to do the average
- and then composite the result with gdalwarp

**Comment by rimds** [comment]
is it not possible to do the same using gdalvrt?

**Comment by mdsumner** [comment]
this would be amazing, I'll try the calc approach I hadn't thought of doing it that way

**Comment by jratike80** [comment]
I believe this is now supported in `gdal raster mosaic` by using `--pixel function mean`
https://gdal.org/en/stable/programs/gdal_raster_mosaic.html#cmdoption-gdal_raster_mosaic-pixel-function
https://gdal.org/en/stable/drivers/raster/vrt.html#builtin-pixel-functions

> (GDAL >= 3.12) Arithmetic mean of input raster bands.
If the optional propagateNoData parameter is set to true, then
if a NoData pixel is found in one of the bands, if will be propagated to
the output value. Otherwise, NoData pixels will be ignored.
